common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -574,14 +574,14 @@ public UTF8String trim() {
   public UTF8String trimAll() {
     int s = 0;
     // skip all of the whitespaces (<=0x20) in the left side
-    while (s < this.numBytes && Character.isWhitespace(getByte(s))) s++;
+    while (s < this.numBytes && getByte(s) <= 0x20) s++;
```

Comment:
Is either behavior more 'correct'? I'm not sure what this is trying to match. It's possible `.isWhitespace` is a better behavior, in which case the docs should change. By default I'd not change behavior unless there's a reason it's not intended.

Reply:
If we comply with the comment for `trimAll`, `<= 0x20` one seems correct. Here is the implementation of `String.trim`.
https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/share/classes/java/lang/StringLatin1.java#L531-L542
https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/share/classes/java/lang/StringUTF16.java#L847-L860

Also, I noticed that originally, characters <= 0x20 were trimmed but #29375 changed the behavior.
That change seems to break the compatibility.
`sql-migration-guide.md` says like as follows.
```
In Spark 3.0, when casting string value to integral types(tinyint, smallint, int and bigint), datetime types(date, timestamp and interval) and boolean type, the leading and trailing whitespaces (<= ASCII 32) will be trimmed before converted to these type values, for example, `cast(' 1\t' as int)` results `1`, `cast(' 1\t' as boolean)` results `true`, `cast('2019-10-10\t as date)` results the date value `2019-10-10`. In Spark version 2.4 and below, when casting string to integrals and booleans, it does not trim the whitespaces from both ends; the foregoing results is `null`, while to datetimes, only the trailing spaces (= ASCII 32) are removed.
```

In fact,  select `cast('2019-10-10\b' as date);` returns `2019-10-10` in `Spark 3.0.0`. 
But after `3.0.1`, the query returns `NULL.

Reply:
Looking at https://github.com/apache/spark/pull/29375 , it seems like the change was at least partly on purpose to catch 'whitespace' that isn't ASCII 32 or less. @WangGuangxin is this change of behavior necessary? Do we need to check for `.isWhitespace` _or_ `<= 0x20`?

Reply:
> Looking at #29375 , it seems like the change was at least partly on purpose to catch 'whitespace' that isn't ASCII 32 or less

I think, the purpose of that change was to handle code points which is >= 0x80 (non-ASCII).
For example, `あ` is `00 81 82` in hex in `UTF-8`.
Before #29375, `getByte` returns `-127` for `0x81` so only checking `<= 0x20` is not enough.
I think this is the problem #29375 originally aimed to resolve.

But it should have checked whether a `byte` data is in the range of `0` and `0x20` to avoid breaking the compatibility.

Reply:
cc: @cloud-fan and @yaooqinn  who were involved in #29375 and #26622.